// eslint-watch/build/options.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _optionator = _interopRequireDefault(require("optionator"));

var _lodash = _interopRequireDefault(require("lodash"));

var _help = _interopRequireDefault(require("./eslint/help"));

var _logger = _interopRequireDefault(require("./logger"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const logger = (0, _logger.default)('options');
logger.debug('Loaded');
const settings = {
  prepend: 'esw [options] [file.js ...] [dir ...]',
  concatRepeatedArrays: true,
  mergeRepeatedObjects: true
};
const myOptions = [{
  heading: 'Options'
}, {
  option: 'help',
  alias: 'h',
  type: 'Boolean',
  description: 'Show help'
}, {
  option: 'format',
  alias: 'f',
  type: 'String',
  default: 'simple-detail',
  description: 'Use a specific output format'
}, {
  option: 'watch',
  alias: 'w',
  type: 'Boolean',
  description: 'Enable file watch'
}, {
  option: 'changed',
  type: 'Boolean',
  description: 'Enables single file linting while watch is enabled'
}, {
  option: 'clear',
  type: 'Boolean',
  description: 'Clear terminal when running lint'
}, {
  option: 'esw-version',
  type: 'Boolean',
  description: 'Prints Eslint-Watch\'s Version'
}];
const eslintOptions = (0, _help.default)();

const newOptions = _lodash.default.union(myOptions, eslintOptions);

settings.options = newOptions;

var _default = (0, _optionator.default)(settings);

exports.default = _default;
module.exports = exports["default"];
